src/Data/FileSystem.elm

```diff
@@ -0,0 +1,186 @@
+module Data.FileSystem exposing
+    ( FileSystem(..)
+    , FolderName
+    , filterFiles
+    , flatten
+    , mapFiles
+    , mkFileSystem
+    , reverse
+    , sortAudienceFilesAndFoldersAlphabetically
+    , sortFoldersAlphabetically
+    , sortWith
+    , toList
+    )
+
+import Data.Audience exposing (Audience)
+import Data.AudienceFolder exposing (AudienceFolder)
+import Data.List.Utility exposing (splitFilter)
+
+
+type alias FolderName =
+    String
+
+
+type FileSystem file
+    = File file
+    | Folder FolderName (List (FileSystem file))
+
+
+toList : FileSystem file -> List file
+toList tree =
+    case tree of
+        Folder name content ->
+            List.concatMap toList content
+
+        File file ->
+            [ file ]
+
+
+{-| Data.FileSystem.flatten
+
+merges all files into the root folder
+
+-}
+flatten : FileSystem file -> FileSystem file
+flatten tree =
+    case tree of
+        File _ ->
+            tree
+
+        Folder name contents ->
+            Folder name (List.map File <| toList tree)
+
+
+filterFiles : (a -> Bool) -> FileSystem a -> FileSystem a
+filterFiles check =
+    mapFolders <|
+        List.filter
+            (\ft ->
+                case ft of
+                    File a ->
+                        check a
+
+                    Folder _ _ ->
+                        True
+            )
+
+
+mapFiles : (a -> b) -> FileSystem a -> FileSystem b
+mapFiles f tree =
+    case tree of
+        File x ->
+            File (f x)
+
+        Folder name content ->
+            Folder name (List.map (mapFiles f) content)
+
+
+mapFolders : (List (FileSystem a) -> List (FileSystem a)) -> FileSystem a -> FileSystem a
```

Comment:
this is functor probably but in a different category than one of set and functions. I would probably use `update` name as it's `a -> a`